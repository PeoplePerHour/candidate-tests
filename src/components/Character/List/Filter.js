import React, { PureComponent } from 'react';
import cn from 'classnames';

import Downshift from 'downshift';
import { Button } from '../../Common';
import './Filter.scss';

class Filter extends PureComponent {
  onChange = selected => {
    this.props.onChange(selected);
  };

  render() {
    const { initial, data, withTextInput, inputProps, children } = this.props;

    return (
      <Downshift
        initialSelectedItem={initial}
        onSelect={this.onChange}
        itemToString={item => (item ? item : '')}
      >
        {({
          getToggleButtonProps,
          selectedItem,
          getItemProps,
          isOpen,
          clearSelection
        }) => (
          <div className="Filter">
            <div
              className={cn('Filter__trigger', {
                'Filter__trigger--active':
                  !!selectedItem || (inputProps && !!inputProps.value)
              })}
            >
              <Button {...getToggleButtonProps()} plain>
                <span>{children}</span>
                <span style={{ marginLeft: 20 }}>⯆</span>
              </Button>
            </div>
            {!isOpen ? null : (
              <div className={cn('Filter__items', { open: isOpen })}>
                {withTextInput && <input type="text" {...inputProps} />}
                <div className="Filter__items-options">
                  {data.map((item, index) => (
                    <Button
                      key={item}
                      plain
                      active={selectedItem === item}
                      id={item}
                      value={item}
                      style={{
                        textAlign: 'left'
                      }}
                      {...getItemProps({
                        item,
                        index
                      })}
                      onClick={e => {
                        if (selectedItem === item) {
                          clearSelection(e);
                        } else {
                          getItemProps({ item, index }).onClick(e);
                        }
                      }}
                    >
                      <span
                        style={{
                          display: 'flex',
                          justifyContent: 'space-between'
                        }}
                      >
                        <span>{item}</span>
                        <span>{selectedItem === item ? '✅' : ''}</span>
                      </span>
                    </Button>
                  ))}
                </div>
                <Button plain onClick={clearSelection}>
                  clear filter
                </Button>
              </div>
            )}
          </div>
        )}
      </Downshift>
    );
  }
}

export default Filter;
