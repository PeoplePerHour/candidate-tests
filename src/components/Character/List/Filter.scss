@import '~styles/colors';

.Filter {
  position: relative;
}

.Filter__items {
  position: absolute;
  display: flex;
  flex-direction: column;
  top: 100%;
  left: 0;
  background-color: white;
  min-width: 200px;
  background-clip: padding-box;
  border: 1px solid $gray-light;
  box-shadow: 1px 1px 10px rgba(0, 0, 0, 0.05);

  > .Button {
    padding: 5px 5px 10px;
    font-size: 12px;
    color: $gray-dark;
  }
}

.Filter__items-options {
  overflow: auto;
  max-height: 300px;
  display: flex;
  flex-direction: column;

  .Button {
    &:hover {
      background-color: $gray-light;
    }
  }
}

.Filter__trigger {
  position: relative;
  border-right: 1px solid $gray-light;
  &--active {
    &:before {
      border-color: $orange transparent transparent $orange;
      border-style: solid;
      border-width: 10px;
      content: '';
      position: absolute;
    }
  }
}

.Filter {
  input[type="text"] {
    margin: 10px 20px 5px;
    padding: 5px;

    &:focus {
      border-color: $theme-color;
      border-shadow: 0;
    }
  }
}
