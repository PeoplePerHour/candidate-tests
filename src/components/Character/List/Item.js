import React, { PureComponent } from 'react';
import { connect } from 'react-redux';
import { openCharacterModal } from '../../../actions';

import { Avatar } from '../../Common';
import Status from '../Status';

import './Item.scss';

class CharacterListItem extends PureComponent {
  openModal = () => {
    const { id, openCharacterModal } = this.props;
    openCharacterModal({ id });
  };

  render() {
    const {
      image,
      origin,
      name,
      status,
      species,
      subspecies,
      gender,
      episode,
      location
    } = this.props;

    return (
      <div role="button" onClick={this.openModal} style={{ cursor: 'pointer' }}>
        <div className="CharacterListItem">
          <div className="CharacterListItem__Avatar">
            <Avatar src={image} alt={name} />
          </div>
          <div className="CharacterListItem__Info">
            <header>
              <Status text={status} />
              <h5>
                {name}
                <span>
                  from <b>{origin.name}</b>
                </span>
              </h5>
            </header>
            {gender}, {species}
            {subspecies && `/${subspecies}`}
            <br />
            {location.name}
            <br />
            Episodes: {episode.length}
          </div>
        </div>
      </div>
    );
  }
}

export default connect(
  null,
  { openCharacterModal }
)(CharacterListItem);
