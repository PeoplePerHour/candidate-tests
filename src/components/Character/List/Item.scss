@import '~styles/colors';

.CharacterListItem {
  display: flex;
  align-items: flex-start;
  background-color: white;
  padding: 15px;
  position: relative;
  z-index: 1;

  &:not(:last-child) {
    border-bottom: 1px solid $gray-light;
  }

  &:hover {
    transform: scale(1.01);
    box-shadow: 1px 1px 10px rgba(0,0,0,.05);
    z-index: 2;
  }
}

.CharacterListItem__Avatar {
  width: 120px;
  height: 120px;
}

.CharacterListItem__Info {
  display: flex;
  align-items: flex-start;
  flex-direction: column;
  padding-left: 15px;
  width: 100%;

  > header {
    display: inline-flex;
    flex-wrap: wrap;
    align-items: center;
    margin-bottom: 2px;

  }

  h5 {
    font-size: 20px;
    margin: 0 0 0 5px;

    > a {
      text-decoration: none;
      color: black;
    }

    > span {
      margin: 0 0 0 5px;
      color: $gray-dark;
      font-size: 14px;
    }
  }
}
