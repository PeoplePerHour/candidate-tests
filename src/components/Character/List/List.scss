.CharacterList {
  max-width: 1200px;
  margin: 0 auto;
  padding: 20px;
}

.CharacterList__items-container {
  box-shadow: 1px 1px 10px rgba(0,0,0,.05);
  position: relative;
}

.CharacterList__footer {
  display: flex;
  justify-content: space-between;
}

.CharacterList__loading {
  position: absolute;
  z-index: 2;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-color: rgba(white, .7);
}
