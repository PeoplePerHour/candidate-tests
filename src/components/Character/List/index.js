export { default } from './List';
export { default as CharacterListItem } from './Item';
