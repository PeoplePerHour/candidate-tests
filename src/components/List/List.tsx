import * as React from "react";
import * as actions from "../../redux/actions";
import { connect } from "react-redux";
import ListItem from "../ListItem/ListItem";
import Modal from "../Modal/Modal";
import Details from "../Details/Details";
import PagerButton from "../PagerButton/PagerButton";
import { PropsWithChildren } from "react";
import * as models from "../../models";
import DropDown from "../DropDown/DropDown";

const nothing = "/src/assets/nothing.png";

const status = ["alive", "dead", "unknown"];
const gender = ["female", "male", "genderless", "unknown"];

interface Iprops {
  fetchCharacters: any;
  characters: any;
  species: any;
  info: any;
  loading: any;
  location: any;
  history: any;
  errorInSearch: boolean;
}

const defaultValues: any = {
  species: null,
  status: null,
  gender: null
};

const List: React.FC = (props: PropsWithChildren<Iprops>) => {
  const {
    fetchCharacters,
    characters,
    species,
    info,
    loading,
    location,
    history,
    errorInSearch
  } = props;

  const [drodownValues, setDefaultValues] = React.useState(defaultValues);
  const [modalOpen, setModal] = React.useState(false);
  const [selectedCharacter, setCharacter] = React.useState(null);

  React.useEffect(() => {
    fetchCharacters(location ? location.search : "");
    getDefaultValue();
  }, [location]);

  const getDefaultValue = () => {
    let params = new URLSearchParams(document.location.search.substring(1));
    setDefaultValues({
      species: params.get("species"),
      status: params.get("status"),
      gender: params.get("gender")
    });
  };

  const HandleModal = (data: models.ICharacter[]) => {
    setModal(!modalOpen);
    setCharacter(data);
  };

  const handleLocationChangeChange = (el: string, val: any) => {
    let params = new URLSearchParams(document.location.search.substring(1));
    if (
      !val ||
      val == "Select Species" ||
      val == "Select Status" ||
      val == "Select Gender"
    ) {
      params.delete(el);
    } else {
      params.set(el, val);
    }
    history.push({
      pathname: location.pathname,
      search: params.toString()
    });
  };

  const GoBack: React.FC = () => {
    return (
      <div className={"nothing-here d-flex fd-col fa-center p404"}>
        <img src={nothing} />
        <h1>There is nothing Here !</h1>
        <button className="btn" onClick={() => history.goBack()}>
          Go Back!
        </button>
      </div>
    );
  };

  return (
    <div>
      <div className={"d-flex"}></div>

      <h1 className="title">Rick and Morty</h1>
      {loading && errorInSearch ? (
        <GoBack />
      ) : (
        <div className={"list-container d-flex fd-col fa-center"}>
          <div className={"d-flex controls "}>
            <PagerButton
              handleNavigate={handleLocationChangeChange}
              text={"Prev"}
              querySting={history}
              link={info.prev}
            />
            <DropDown
              Options={species}
              onChange={val => handleLocationChangeChange("species", val)}
              value={drodownValues.species ? drodownValues.species : null}
              defaultText={"Select Species"}
            />
            <DropDown
              Options={status}
              onChange={val => handleLocationChangeChange("status", val)}
              value={drodownValues.status ? drodownValues.status : null}
              defaultText={"Select Status"}
            />
            <DropDown
              Options={gender}
              onChange={val => handleLocationChangeChange("gender", val)}
              value={drodownValues.gender ? drodownValues.gender : null}
              defaultText={"Select Gender"}
            />

            <PagerButton
              handleNavigate={handleLocationChangeChange}
              text={"Next"}
              querySting={history}
              link={info.next}
            />
          </div>
          <div>
            <ul className={"d-flex fa-center m-auto list"}>
              {characters.map((character: models.ICharacter) => {
                return (
                  <ListItem
                    click={HandleModal}
                    key={character.id}
                    data={character}
                  />
                );
              })}
            </ul>
          </div>
        </div>
      )}
      <Modal displayModal={modalOpen} closeModal={HandleModal}>
        <Details character={selectedCharacter} />
      </Modal>
    </div>
  );
};

const mapStateToProps = (state: models.IData) => {
  return {
    characters: state.data.characters,
    info: state.data.info,
    loading: state.data.loading,
    species: state.data.species,
    errorInSearch: state.data.errorInSearch
  };
};

const mapDispatchToStore = (dispatch: any) => {
  return {
    fetchCharacters: (query: string) => dispatch(actions.fetchCharacters(query))
  };
};

export default connect(mapStateToProps, mapDispatchToStore)(List);
