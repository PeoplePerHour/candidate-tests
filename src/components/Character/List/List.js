import React, { Component } from 'react';
import qs from 'qs';

import { connect } from 'react-redux';
import { fetchCharacters } from '../../../actions';
import {
  getCharacters,
  getCharacterPaginationInfo,
  getCharactersIsFetching
} from '../../../selectors';

import CharacterListItem from './Item';
import CharacterListFooter from './ListFooter';
import CharacterListHeader from './ListHeader';
import './List.scss';

export class CharacterList extends Component {
  constructor(props) {
    super(props);
    this.loadData = this.loadData.bind(this);
    this.onFilter = this.onFilter.bind(this);
  }

  componentDidMount() {
    this.loadData();
  }

  loadData(params = this.props.params) {
    this.props.fetchCharacters(params);
  }

  onFilter({ gender, species, status }) {
    const { history } = this.props;
    this.loadData({ gender, species, status });
    history.push({
      pathname: history.location.pathname,
      search: qs.stringify({ gender, species, status })
    });
  }

  render() {
    const { characters, info, loading, params } = this.props;

    return (
      <div className="CharacterList">
        <CharacterListHeader params={params} onFilter={this.onFilter} />

        <div className="CharacterList__items-container">
          {loading && <div className="CharacterList__loading" />}
          {characters.map(n => (
            <CharacterListItem key={n.id} {...n} />
          ))}
        </div>
        <CharacterListFooter onClick={this.loadData} info={info} />
      </div>
    );
  }
}

const mapStateToProps = (state, ownProps) => ({
  characters: getCharacters(state, ownProps.params),
  info: getCharacterPaginationInfo(state),
  loading: getCharactersIsFetching(state),
  params: qs.parse(ownProps.history.location.search, {
    ignoreQueryPrefix: true
  })
});

export default connect(
  mapStateToProps,
  {
    fetchCharacters,
    getCharacterPaginationInfo
  }
)(CharacterList);
