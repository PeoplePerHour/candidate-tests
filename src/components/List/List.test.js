
import * as  React from 'react';
import * as ReactDOM from 'react-dom';
import { Provider } from "react-redux";
import store from "../../redux/store";



import List from './list'



it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Provider store={store}><List/></Provider>, div);
    expect(div).toMatchSnapshot();
    ReactDOM.unmountComponentAtNode(div);
})
