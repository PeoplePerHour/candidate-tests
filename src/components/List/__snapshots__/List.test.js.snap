// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div>
  <div>
    <div
      class="d-flex"
    />
    <h1
      class="title"
    >
      Rick and Morty
    </h1>
    <div
      class="list-container d-flex fd-col fa-center"
    >
      <div
        class="d-flex controls "
      >
        <button
          class="btn"
        >
          Prev
        </button>
        <div
          class="select-dd"
        >
          <span>
            Select Species
          </span>
          <ul
            class="hide"
          >
            <li>
              <a
                href="#"
              >
                Select Species
              </a>
            </li>
          </ul>
        </div>
        <div
          class="select-dd"
        >
          <span>
            Select Status
          </span>
          <ul
            class="hide"
          >
            <li>
              <a
                href="#"
              >
                Select Status
              </a>
            </li>
            <li>
              <a
                href="#"
              >
                alive
              </a>
            </li>
            <li>
              <a
                href="#"
              >
                dead
              </a>
            </li>
            <li>
              <a
                href="#"
              >
                unknown
              </a>
            </li>
          </ul>
        </div>
        <div
          class="select-dd"
        >
          <span>
            Select Gender
          </span>
          <ul
            class="hide"
          >
            <li>
              <a
                href="#"
              >
                Select Gender
              </a>
            </li>
            <li>
              <a
                href="#"
              >
                female
              </a>
            </li>
            <li>
              <a
                href="#"
              >
                male
              </a>
            </li>
            <li>
              <a
                href="#"
              >
                genderless
              </a>
            </li>
            <li>
              <a
                href="#"
              >
                unknown
              </a>
            </li>
          </ul>
        </div>
        <button
          class="btn"
        >
          Next
        </button>
      </div>
      <div>
        <ul
          class="d-flex fa-center m-auto list"
        />
      </div>
    </div>
    <div
      class="modal"
      style="display: none;"
    >
      <div
        class="modal-content"
      >
        <span
          class="close"
        >
          Ã—
        </span>
        <div>
          <div
            class="modal-flex"
          />
        </div>
      </div>
      <div
        class="modal-background"
      />
    </div>
  </div>
</div>
`;
